### PR TITLE
4341 implement a new version of the autocomplete which features suggestions and synonym matching

### PR DESCRIPTION
**Pull request by tomas-stefano**
## Context

In order to make it clearer to candidates that they don't HAVE to select an option from the degree subject or institution autocompletes, but that they can type in another option, we've designed a tweak to the autocomplete.

The new version has:

* Search in the synonyms (or other fields like abbreviation)
* A gap immediately beneath the input, so it looks less like a dropdown
* A "Suggestions" header
* Does not show anything if no results match so that it looks like a regular input
* Hint text which explicitly says that users can select from suggestions or enter their own

## Guidance to review

1. Is the autocomplete for degree types, degree subjects, institutions and degree grades working?
2. Does it show the suggestion header?
3. Does it show the hint text?
4. Is the synonyms working with the [new data](https://github.com/DFE-Digital/dfe-reference-data/blob/main/lib/dfe/reference_data/degrees.rb)?

## Link to Trello card

https://trello.com/c/yytMJzYs/4341-implement-a-new-version-of-the-autocomplete-which-features-suggestions-and-synonym-matching

### CODE REVIEW

**Comment by duncanjbrown** [comment]
Nit: of these three, only the first 2 have results in the autocomplete

Screenshot 2022-04-13 at 10 10 07

File: app/controllers/candidate_interface/degrees/type_controller.rb [hunk 1]
```diff
@@ -43,15 +43,15 @@ def update
...
 
       def set_degree_type_names
-        @degree_types = Hesa::DegreeType.abbreviations_and_names.sort
+        @degree_types = Hesa::DegreeType.all
       end
 
       def conditionally_render_new_degree_type_form
         if degree_already_added?
-          @degree_types = Hesa::DegreeType.abbreviations_and_names
+          set_degree_type_names
```
Comment (duncanjbrown) [inline app/controllers/candidate_interface/degrees/type_controller.rb:51]: Minor: would prefer to set the ivar inline here, as in the alternative branch of the if?

File: app/views/candidate_interface/degrees/grade/_form_fields.html.erb [hunk 2]
```diff
@@ -23,7 +23,7 @@
       <%= f.govuk_radio_button :grade, grade, label: { text: grade }, link_errors: i.zero? %>
     <% end %>
     <%= f.govuk_radio_button :grade, 'other', label: { text: t('application_form.degree.grade.other.label') } do %>
-      <%= render Dfe::Autocomplete::View.new(
+      <%= render DfE::Autocomplete::View.new(
```
Comment (duncanjbrown) [inline app/views/candidate_interface/degrees/grade/_form_fields.html.erb:26]: 👌

File: spec/support/test_helpers/candidate_helper.rb [hunk 3]
```diff
@@ -305,13 +305,13 @@ def and_the_candidate_add_the_degree(degree_type:, degree_subject:, institution:
     visit candidate_interface_new_degree_path
 
     choose 'UK degree'
-    fill_in 'Type of degree', with: degree_type
+    select degree_type, from: 'Type of degree'
```
Comment (duncanjbrown) [inline spec/support/test_helpers/candidate_helper.rb:308]: Should we have at least one `js: true` spec for this? (not in this file!)

**Review by duncanjbrown — COMMENTED** [review]
Looking good! I didn't quite understand https://github.com/DFE-Digital/apply-for-teacher-training/pull/6784/commits/b277a6f9f7c7d0e7170eb1b09f84619305b0bead?

**Comment by tomas-stefano** [comment]
> Looking good! I didn't quite understand [b277a6f](https://github.com/DFE-Digital/apply-for-teacher-training/commit/b277a6f9f7c7d0e7170eb1b09f84619305b0bead)?

Yes, it was difficult for me to understand as well.

Before this commit above, the test was creating a degree in the database and the test was editing the degree's info, one by one.

The problem is that the user edits the degree and changes the degree type the grade was set to nil, therefore the assertion for the grade to be checked was failing. [There is an implementation that sets the grade to nil if there is no grade HESA code or the grade is not the main undergrad](https://github.com/DFE-Digital/apply-for-teacher-training/blob/main/app/forms/candidate_interface/degree_type_form.rb#L34) so the assertion that the grade that was set up at the beginning of the specs was being overwritten by the step where the user changes the degree type.
The solution was to set the grade hesa code at the same time set the main undergrad grade in the beginning of the spec.

**Review by duncanjbrown — APPROVED** [review]
💯 Thanks for adding spec. This works really well!

**Pull request merged**